### ISSUE TITLE
Multiple passwords presented despite only one matching URL

### ISSUE THREAD

**Issue by vbeffa** [keepassxreboot/keepassxc-browser#2164]
I have two passwords set up for my ASUS router. For one the URL is `http://192.168.1.1/Main_Login.asp`. For the other the URL is `http://192.168.1.1/Advanced_OpenVPNClient_Content.asp`. On `http://192.168.1.1/Advanced_OpenVPNClient_Content.asp`, the second password is correctly auto-filled. On `http://192.168.1.1/Main_Login.asp`, both passwords are presented, and neither is auto-filled.

## Expected Behavior
Only one password is presented and auto-filled on `http://192.168.1.1/Main_Login.asp`.

## Current Behavior
Two passwords are presented on `http://192.168.1.1/Main_Login.asp`.

## Steps to Reproduce (for bugs)
1. Create router passwords as described above.
2. Navigate to the URLs above.

## Debug info

KeePassXC - 2.7.7
KeePassXC-Browser - 1.9.0.2
Operating system: Mac
Browser: Firefox

**Comment by varjolintu** [comment]
This is expected, because at this point only the domain (here's the IP address) only matters if no paths are used.

**Comment by vbeffa** [comment]
Sorry I don't follow - both URLs use a path, and the second one (`http://192.168.1.1/Advanced_OpenVPNClient_Content.asp`) correctly auto-fills the username and password.

**Comment by varjolintu** [comment]
That entry URL will match everything under `http://192.168.1.1/*`. If you use paths in the URL and in your local server, the matching will work correctly. Meaning `https://192.168.1.1/path` will not be matched with `https://192.168.1.1/secondpath`.

**Comment by vbeffa** [comment]
I'm still not following. I'm using paths in the URLs and my browser:

- `http://192.168.1.1/Main_Login.asp` - path is `/Main_Login.asp`
- `http://192.168.1.1/Advanced_OpenVPNClient_Content.asp` - path is `/Advanced_OpenVPNClient_Content.asp`

For both, I use the full path in the password entry (**NOT** just `http://192.168.1.1`)  and when I navigate to the pages in my browser. The first entry does not auto-fill on the login page but the second does on the OpenVPN page.

**Comment by varjolintu** [comment]
What I mean there's a difference between:
`http://192.168.1.1/main/Main_Login.asp` and `http://192.168.1.1/anotherAdvanced_OpenVPNClient_Content.asp`

But if you serve both files under `http://192.168.1.1` there's no difference and both entries are returned.
Defining an exact URL is done here for KeePassXC 2.8.0: https://github.com/keepassxreboot/keepassxc/pull/9835

**Comment by vbeffa** [comment]
Are you referring to having a folder between the domain and the file? Again, how does that explain why the second URL auto-fills correctly? The PR you linked is in draft state - what am I supposed to do with it?

**Comment by varjolintu** [comment]
Are you using `http://192.168.1.1` as your entry URL?

The PR I linked just addresses an issue about exact URL's, if you want certain entries to specify them.

**Comment by vbeffa** [comment]
No, I'm using the full paths as I stated. (Unfortunately KeePass seems to somehow be blocking me from taking screenshots of it. Apparently I'm not intelligent enough to not take an image of my password and only the URL, so Keepass needs to block me from all screenshots. I don't see a setting to disable this, so I'm unable to show the URLs other than as stated in my earlier comments.) The PR is not merged, so whatever changes it is making are not available in the downloaded version I am using, so what changes should I make in my entries?

**Comment by varjolintu** [comment]
Ok, I understand. I'll take a look at the issue and try to reproduce it meyself.

**Comment by vbeffa** [comment]
OK thank you.

**Comment by BloodyRain2k** [comment]
> That entry URL will match everything under `http://192.168.1.1/*`. If you use paths in the URL and in your local server, the matching will work correctly. Meaning `https://192.168.1.1/path` will not be matched with `https://192.168.1.1/secondpath`.

I wish that's how it would work: I have two login's stored for a KeyCloak instance I manage, my normal login for the realm and the administrative login for KeyCloak's own realm.

Url-wise they differ like this: `https://example.com/auth/realms/master/protocol/openid-connect/auth?...` for the administration and `https://example.com/auth/realms/project/protocol/openid-connect/auth?...` for the normal usage.

But the extension shows both logins for both urls, even though I have configured the url's of each entry for `https://example.com/auth/realms/master/` and `https://example.com/auth/realms/project/` respectively.

I've also had the issue described by varjolintu with my Synology NAS, which does it in an even more annoying way by putting the login "path" behind the `#` of the actual url. So it looks like `https://192.168.178.44:5001/#/signin`.
Of course that doesn't work either (it sees the entire NAS as a potential login form), but here I suspected that it was just the `#` that broke it, but the case with KeyCloak suggests it really doesn't care beyond the domain for filtering which logins to show.

**Comment by varjolintu** [comment]
@BloodyRain2k Maybe using exact URLs could help: https://github.com/keepassxreboot/keepassxc/pull/9835. That only works with Additional URLs though.

**Comment by DragRedSim** [comment]
I can confirm this is still an issue.

I have a setup with a Home Assistant instance (in this case, let's say it's at `https://ha.example.com/`; not a real site). My login for that has a URL field of `https://ha.example.com/auth/authorize`, which is the address of the login page (with a bunch of automatic arguments following a question mark added in the redirect from the root address). Within that instance, I run the Portainer addon, which requires a separate username and password combination. I have this with a URL field of `https://ha.example.com/db21ed7f_portainer`, which is the displayed URL on that page when accessed through the HA interface, along with a secondary URL of `https://ha.example.com/api/hassio_ingress/(install-specific URL tag)`, found by getting the address of the iframe for the Portainer login page. Note that this is not a cross-origin iframe relative to the displayed URL, so the specific setting for this domain should not be necessary, and in my case is disabled (I have the "Improved input field detection" enabled, as the HA login page requires it after their latest update, currently 2026.4.0).

In this situation, both accounts are listed on both the HA login and Portainer login pages; though they share a domain, the HA login should not be visible on the Portainer login page, as it has the `/auth/` subdirectory in the URL path which is not in either the browser's displayed path or the path of the iframe within, while the Portainer login may be considered at the root of the subdomain (since its URL is set to include a substring which does not include an explicit subdirectory), but it should not be displayed when accessing the HA login page at the `/auth/authorize` URL, since neither path for that entry includes the `/auth/` string.

However, I think this might not be the fault of the extension; as far as I can tell, it always sends the full URL of the page that it is looking for passwords for, and it is the KeePassXC application itself which resolves matching entries based on the host only, and not any other elements of the URL path. If so, it would require the application to identify and return only the best match(es), rather than all for the domain, or for the extension to filter the returned entries to the best match(es).

EDIT: Immediately after commenting this, I found the option in the application to "Return only best-matching credentials"... which does exactly what I want in this instance.

**Comment by BloodyRain2k** [comment]
I've been meaning to give an update on this as well, but kept forgetting about it...

For me neither approach works. Including the mentioned `Return only best-matching credentials`.
It keeps showing everything just based on the domain.
And after turning on the option above, it even started to omit the correct option (path and domain match) for a login page entirely and instead returned one that had the wrong path and only matched by domain.

**Comment by varjolintu** [comment]
If you want to use an exact URL for some entry, add the URL to Additional URLs list and put it between `"`, e.g. `"https://example.com/specific_login/page/"`.

**Comment by BloodyRain2k** [comment]
> If you want to use an exact URL for some entry, add the URL to Additional URLs list and put it between `"`, e.g. `"https://example.com/specific_login/page/"`.

That might work for pages with a fully static url, but for the case of KeyCloak, where it's only static up to the `?` this does not seem to work.
Before the `"` I saw every login for the entire domain, regardless of the path after that, and now it's claiming that there are no logins at all.

Likely because `"https://site.com/auth/realms/master/protocol/openid-connect/auth"` does not exactly match `https://site.com/auth/realms/master/protocol/openid-connect/auth?client_id=...` and everything else.

I also tried `"https://site.com/auth/realms/master/protocol/openid-connect/auth?*"` but then KC complained about an invalid URL. So at this point I'm not sure if the issue comes from the app or the extension.
But if the app tends to return all matches by domain, couldn't the extension get an option to then filter that response more by path?


Edit: after reverting my url entries to not have `"` anymore and instead just having them like `https://site.com/auth/realms/master/protocol/openid-connect/auth?*` it suddenly started respecting the path?...
I am very confused, because I swear I tried that in the past and it did NOT work...
I'll continue testing and maybe (hopefully) this sticks, as it would finally be what I wanted.

**Comment by BloodyRain2k** [comment]
I have noticed something that was probably the reason, why the first time I tried the Additional URLs I thought it doesn't work:
If an url is entered __without__ a `*` at the end (somewhere in it?), it appears that the matching still only cares about the domain, regardless of the remaining path of the url.
But once a `*` is present, it switches to full matching.

I just changed some entries, because I noticed that my urls were too precise to deal with some alternative url that KeyCloak uses, so I just trimmed them back to the first separating section of the url and in the process I threw the `*` that I had at the end away.
After that change it started mixing all entries again. (`.../realms/master/` being available on `.../realms/something/...` etc)

So I tried adding the `*` at the end back again (`.../realms/master/*`) and suddenly it was filtering correct again.
But I don't know which of the two is at fault here, the extension or KeePass itself.
